netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -198,6 +295,246 @@ _net_ebpf_sock_addr_update_store_entries()
+    }
+
+    return INETADDR_ISLOOPBACK((SOCKADDR*)&socket_address);
+}
+
+/**
+ * @brief Compare the destination address in the two provided bpf_sock_addr_t structs.
+ *
+ * @param[in] addr1 Pointer to first sock_addr struct to compare.
+ * @param[in] addr2 Pointer to second sock_addr struct to compare.
+
+ * @return true, if the addresses are same, false otherwise.
+ */
+static inline bool
+_net_ebpf_ext_compare_destination_address(_In_ const bpf_sock_addr_t* addr1, _In_ const bpf_sock_addr_t* addr2)
+{
+    ASSERT(addr1->family == addr2->family);
+    if (addr1->family != addr2->family) {
+        return false;
+    }
+
+    if (addr1->family == AF_INET) {
+        return addr1->user_ip4 == addr2->user_ip4;
+    } else {
+        return (memcmp(&addr1->user_ip6[0], &addr2->user_ip6[0], 16) == 0);
+    }
+}
+
+static NTSTATUS
+_net_ebpf_ext_get_connection_context(
+    uint64_t transport_endpoint_handle,
+    _In_ const bpf_sock_addr_t* sock_addr_ctx,
+    _Out_ net_ebpf_extension_connection_context_t** connection_context)
+{
+    NTSTATUS status = STATUS_NOT_FOUND;
+    KIRQL old_irql;
+
+    *connection_context = NULL;
+    old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+
+    LIST_ENTRY* list_entry = _net_ebpf_ext_connect_context_list.Flink;
+    while (list_entry != &_net_ebpf_ext_connect_context_list) {
+        net_ebpf_extension_connection_context_t* entry =
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_connection_context_t, list_entry);
+        if (entry->key.transport_endpoint_handle == transport_endpoint_handle) {
+            // Check if the destination address and port match.
+            if (_net_ebpf_ext_compare_destination_address(&entry->key.destination, sock_addr_ctx) &&
+                entry->key.destination.user_port == sock_addr_ctx->user_port) {
+                // Found matching entry. Update timestamp and move to front of the queue.
+                entry->timestamp = KeQueryInterruptTime() / 10000;
+                RemoveEntryList(&entry->list_entry);
+                InsertHeadList(&_net_ebpf_ext_connect_context_list, &entry->list_entry);
+
+                *connection_context = entry;
+                status = STATUS_SUCCESS;
+                break;
+            }
+        }
+        list_entry = list_entry->Flink;
+    }
+
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
+
+    return status;
+}
+
+static void
+_net_ebpf_ext_delete_connection_context(_In_ _Post_invalid_ net_ebpf_extension_connection_context_t* context)
+{
+    KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+
+    RemoveEntryList(&context->list_entry);
+    ExFreePool(context);
+    _net_ebpf_ext_connect_context_count--;
+
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
+}
+
+static void
+_net_ebpf_ext_purge_lru_contexts_under_lock(bool delete_all)
+{
+    uint64_t expiry_time = KeQueryInterruptTime() / 10000 - EXPIRY_TIME;
```

Comment:
A bunch of places hard code 10000.  Suggest a macro like
```
#define CONVERT_100NS_UNITS_TO_MS(x) ((x) / 10000)
```

Reply:
fixed.